Development

Shamanou opened the pull request:
Added more parsers and updated rdf4j as a response to the bug report by rajaram, the file i got today works with this implementation

Shamanou committed:
Updated prefix loading

- could not fix reported bug because the parsers of rdf4j are unable to parse the provided file.
- errors are propagated to front end now
- more parsers are added extending input compatibility

Shamanou committed:
Added more parsers to file format detection and updates rdf4j

kburger commented:
Before any other comments: codacy has +51 issues. In general that's a good place to start checking for improvements. We don't have a proper setup for javascript yet, so we can (look at but for now) ignore those.

Shamanou commented:
I adjusted codacy settings

Shamanou committed:
Revert "Fixed bug where new projects could not load excisting shared rdf skeleton"

This reverts commit ed7498d92ef7631d49c68ac79d60e6e94626c361.

Shamanou committed:
Fixed bug where new projects could not load excisting shared rdf skeleton

kburger commented on extensions/fairifier-rdf-extension/deri/grefine/rdf/commands/AddPrefixFromFileCommand.java:
```diff
@@ -2,100 +2,118 @@
+                    format = item.getString();
+                    if (format.equals("text/turtle")) {
+                        format = "TTL";
+                    } else if (format.equals("application/rdf+xml")) {
```
These checks are somewhat duplicated down below, I assume this can be replaced with a single [`Rio.getParserFormatForMIMEType`](http://docs.rdf4j.org/javadoc/latest/org/eclipse/rdf4j/rio/Rio.html#getParserFormatForMIMEType-java.lang.String-).

kburger commented on extensions/fairifier-rdf-extension/deri/grefine/rdf/commands/AddPrefixFromFileCommand.java:
```diff
@@ -2,100 +2,118 @@
+            RDFFormat rdfFormat;
+            if (format.equals("TTL")) {
+                rdfFormat = RDFFormat.TURTLE;
+            } else if (format.equals("N3")) {
```
This list of `if`/`else` branches is redundant if the format is determined up above.

kburger commented on extensions/fairifier-rdf-extension/deri/grefine/rdf/commands/AddPrefixFromFileCommand.java:
```diff
@@ -2,100 +2,118 @@
+            Repository repository =
+                    new SailRepository(new ForwardChainingRDFSInferencer(new MemoryStore()));
+            repository.initialize();
+            RepositoryConnection con = repository.getConnection();
```
`RepositoryConnection` implements [`AutoClosable`](http://docs.rdf4j.org/javadoc/latest/org/eclipse/rdf4j/repository/RepositoryConnection.html), so this can be surrounded with a try-with-resources.

kburger commented on extensions/fairifier-rdf-extension/deri/grefine/rdf/commands/AddPrefixFromFileCommand.java:
```diff
@@ -2,100 +2,118 @@
+                    }
+                } else {
+                    filename = item.getName();
+                    in = item.getInputStream();
```
Is there any chance that either:
1. more than one file is uploaded, or
2. other parameters are passed?

If so, this will fail in unexpected ways.

Shamanou replied:
the input dialog only allows for 1 file at a time and there are no other parameters.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -1,145 +1,153 @@
+    RDFParser rdfParserTrix = new TriXParser();
+    RDFParser rdfParserTrig = new TriGParser();
 
     RDFParser[] parsers;
```
1. This is missing an access modifier,
2. Can you explain why you chose an array and not a `Collection` type?

Shamanou replied:
we dont need a collection type, the number of parsers is predifined and doesnt change dynamically we also do not need keys, we do not need to know which parses is which because we can look at the values returned by the RDFParser parent class

kburger replied:
The choice for a collection is not only determined by the size. Ease of interaction, streams, etc.. A map is not a collection, so there are no keys involved.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -1,145 +1,153 @@
+
+        this.parsers = new RDFParser[] {this.rdfParserntriples, this.rdfParserttl,
+                this.rdfParserrdfxml, this.rdfParserJsonld, this.rdfParserNQuads,
+                this.rdfParserJson, this.rdfParserTrix, this.rdfParserTrig};
```
This array will be initialized every single time this command is invoked. Doesn't seem to be the intent? Or at least it should't be.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -1,145 +1,153 @@
         }
+        StringBuffer response = new StringBuffer();;
 
         if ((url != null) && !url.trim().equals("") && !url.trim().equals("url")) {
```
I get that this tries to parse the load and parse the external files. But every single thing in one method is not readable (especially with zero inline comments explaining the flow). This could better be split up into separate methods.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -1,145 +1,153 @@
+            for (int i = 0; i < in.length; i++) {
+                in[i] = new BufferedInputStream(
+                        new ByteArrayInputStream(response.toString().getBytes("UTF8")));
+            }
```
I get that this (entire) block tries to determine the format for an external file, but there are other ways to achieve that. Namely, using a combination of [`RDFFormat.getAcceptParams`](http://docs.rdf4j.org/javadoc/latest/org/eclipse/rdf4j/rio/RDFFormat.html#getAcceptParams-java.lang.Iterable-boolean-org.eclipse.rdf4j.rio.RDFFormat-), [`HttpUrlConnection. addRequestProperty`](http://docs.oracle.com/javase/8/docs/api/java/net/URLConnection.html#addRequestProperty-java.lang.String-java.lang.String-), and [`Rio.getParserFormatForMIMEType`](http://docs.rdf4j.org/javadoc/latest/org/eclipse/rdf4j/rio/Rio.html#getParserFormatForMIMEType-java.lang.String-). Ping me for more details.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -1,145 +1,153 @@
-                parser.parse(in, baseuri);
-                format = parser.getRDFFormat().getDefaultMIMEType();
+                parsers[i].parse(in[i], baseuri);
+                format = parsers[i].getRDFFormat().getDefaultMIMEType();
```
Looping over the known parsers and parsing the file in its entirety should be a last resort instead of the main course of action.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -1,145 +1,153 @@
+            } catch (Exception e) {
+                System.out.println(
+                        parsers[i].getRDFFormat().getDefaultMIMEType() + " : " + e.toString());
                 continue;
```
Using exceptions as main control flow is not preferable. Try to avoid doing that.

Shamanou replied:
This is necessary for this method to work, i know its not preferred but the parsers arise an exception when the input cannot be parsed if this does not happen we know the file format

kburger commented on extensions/fairifier-rdf-extension/scripts/rdf-schema-new-prefix-widget.js:
```diff
@@ -55,7 +55,7 @@ NewPrefixWidget.prototype.show = function(msg,def_prefix,onDone){
                     if(!format['RDFFormat']){
                         dismissBusy();
-                        alert('file format could not be detected, only turtle, rdfxml and ntripple are supported')
+                        alert('file format could not be detected, turtle, rdfxml, ntripple, trig, trix, rdfjson, jsonld and nquads are supported')
```
Typo on N-Triples. The other formats could also use proper capitalization.

kburger commented on main/webapp/modules/core/scripts/project/menu-bar-extensions.js:
```diff
@@ -0,0 +1,519 @@
+	
+	RdfReconciliationManager.synchronizeServices();
+});
+
```
Why is this javascript file marked as new?

Shamanou replied:
This is overwritten by the ant script build.xml

kburger reviewed (changes requested):
There is a lot wrong with this PR. I assume a number of issues are due to the legacy code, but certain choices are new as far as I can tell. Ping me if you need more details.

Shamanou committed:
Adjusted code with pr comments

Shamanou committed:
For now just let travis try building not testing

Shamanou committed:
only test on java 8

kburger commented on extensions/fairifier-rdf-extension/deri/grefine/rdf/commands/AddPrefixFromFileCommand.java:
```diff
@@ -29,91 +39,64 @@ public AddPrefixFromFileCommand(ApplicationContext ctxt) {
+        repository.initialize();
+        try (RepositoryConnection con = repository.getConnection()) {
+            con.add(in, "", format.get());
             con.close();
```
The `con.close()` is redundant when using try-with-resources.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -108,23 +107,31 @@ public void doPost(HttpServletRequest req, HttpServletResponse res)
+
+    private String ParseFullFile(RDFParser[] parsers, InputStream[] in, String baseuri) {
+        String format = null;
         for (int i = 0; i < parsers.length; i++) {
```
You can loop over the array using a foreach loop, makes it less error prone.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -134,21 +141,25 @@ public void doPost(HttpServletRequest req, HttpServletResponse res)
+
+        for (RDFParser parser : parsers) {
+            HttpURLConnection con = (HttpURLConnection) obj.openConnection();
+            con.addRequestProperty("Accept", parser.getRDFFormat().getDefaultMIMEType());
```
`"Accept"` can be a constant, if you have something like guava or spring accessible then use their `HttpHeaders` objects.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -1,146 +1,165 @@
+
+        for (RDFParser parser : parsers) {
+            HttpURLConnection con = (HttpURLConnection) obj.openConnection();
+            con.addRequestProperty("Accept", parser.getRDFFormat().getDefaultMIMEType());
```
There's no need to loop over the parsers one by one, you can generate the accept header values as I said [here](https://github.com/DTL-FAIRData/FAIRifier/pull/47#discussion_r145086937), that makes it quite a bit faster (especially for slow servers).

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -1,146 +1,165 @@
+                return parser.getRDFFormat().getDefaultMIMEType();
+            }
+        }
+        return "application/rdf+xml";
```
Do we always want to default to rdfxml, and not to turtle?

Shamanou replied:
i thought rdfxml was more common for ontologies

kburger reviewed (changes requested):
Still room for some improvements.

Shamanou committed:
Added guava and updated code with comments

Shamanou committed:
remove unescasary file

Shamanou committed:
code cleanup

Shamanou committed:
code cleanup

Shamanou committed:
code cleanup

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -1,146 +1,161 @@
+    private static final RDFParser rdfParserTrig = new TriGParser();
+
+    private final static RDFParser[] parsers = { rdfParserntriples, rdfParserttl, rdfParserrdfxml, rdfParserJsonld,
+            rdfParserNQuads, rdfParserJson, rdfParserTrix, rdfParserTrig };
```
```
private static final RDFParser[] PARSERS = {
new TurtleParser(),
new ...
};
```

```
private static final RDFParsers[] PARSERS;

static {
  PARSERS = new RDFParser[] {
    new TurtleParser(),
    ...
  };
}
```

Shamanou committed:
Improved code style in addprefixfromfilecommand and detectfileformatcommand

- Removed uncescesary variable from addprefixfromfilecommand

- moved parsers to arraylist

- added accept headers to get format from url

- removed unescessary jar file depedencies and removed guava

Shamanou commented:
Removed a lot of unneeded jars that is why the files changed is quite hight

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -48,24 +51,27 @@
-
-    private final static RDFParser[] parsers = { rdfParserntriples, rdfParserttl, rdfParserrdfxml, rdfParserJsonld,
-            rdfParserNQuads, rdfParserJson, rdfParserTrix, rdfParserTrig };
+    private final static ArrayList<RDFParser> parsers = new ArrayList<RDFParser>() {
```
This is an unconventional way of initializing the arraylist. Better to create a static initialization block and add it in there.

kburger replied:
The variable `parsers` is static and (should be) immutable. Therefor, convention dictates to rename this to `PARSERS`.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -83,8 +89,7 @@ public void doPost(HttpServletRequest req, HttpServletResponse res)
         String baseuri = null;
-        String filename = null;
-        InputStream[] in = new InputStream[parsers.length];
+        InputStream[] in = new InputStream[parsers.size()];
```
Not sure if we still need this array approach, but we can leave it for now.

Shamanou replied:
We need it because if we parse full files the streams might be closed on an exception

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -130,15 +134,15 @@ public void doPost(HttpServletRequest req, HttpServletResponse res)
+    private String parseFullFile(ArrayList<RDFParser> parsers, InputStream[] in, String baseuri) {
         String format = null;
-        for (int i = 0; i < parsers.length; i++) {
+        for (int i = 0; i < parsers.size(); i++) {
```
Better to use a `foreach` to iterate over the parsers.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -130,15 +134,15 @@ public void doPost(HttpServletRequest req, HttpServletResponse res)
                 }
-                parsers[i].parse(in[i], baseuri);
-                format = parsers[i].getRDFFormat().getDefaultMIMEType();
+                parsers.get(i).parse(in[i], baseuri);
```
When using the `foreach` you don't need to `get()` each parser.

kburger commented on extensions/fairifier-rdf-extension/dtls/fairifier/DetectFileFormatCommand.java:
```diff
@@ -147,11 +151,17 @@ private String ParseFullFile(RDFParser[] parsers, InputStream[] in, String baseu
         StringBuffer response = new StringBuffer();
         URL obj = new URL(url);
         HttpURLConnection con = (HttpURLConnection) obj.openConnection();
+        List<RDFFormat> rdfFormats = parsers.stream().map(val -> val.getRDFFormat()).collect(Collectors.toList());
```
You can get the formats more easily by doing [`RDFParserRegistry.getInstance().getKeys()`](http://docs.rdf4j.org/javadoc/latest/org/eclipse/rdf4j/common/lang/service/ServiceRegistry.html#getKeys--).

kburger reviewed (changes requested):
Small changes, the rest looks nice.

Shamanou committed:
Improved code readability

- Added PARSERS to more commonly used static block

- Made exceptions less generic

- only call stream when it is needed, pass the fileitem handler to private method

kburger reviewed (approved):
Good job, looks fine now.
